unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+                    for link in links {
+                        pending.push(link?);
+                    }
+                    pending
+                };
+
+                let state = if let Some(next) = pending.pop() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next: Some(next),
+                        pending,
+                    })
+                } else {
+                    State::Last(self.current)
+                };
+
+                Ok(ContinuedWalk::Directory(Item::from(state)))
+            },
+            UnixFsType::HAMTShard => {
+                let (cid, name, depth) = self.next.expect("continued without next");
+                self.current.as_bucket(cid, &name, depth);
+
+                // similar to directory the depth is +1 for nested entries, but the sibling buckets
+                // are at depth
+                let mut links = flat.links
+                    .into_iter()
+                    .enumerate()
+                    .map(|(nth, link)| convert_sharded_link(depth + 1, nth, link))
+                    .rev();
+
+                let mut pending = {
+                    let mut pending = self.pending;
+                    for link in links {
+                        pending.push(link?);
+                    }
+                    pending
+                };
+
+                let state = if let Some(next) = pending.pop() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next: Some(next),
+                        pending,
+                    })
+                } else {
+                    State::Last(self.current)
+                };
+
+                Ok(ContinuedWalk::Directory(Item::from(state)))
+
+            },
+            UnixFsType::Raw | UnixFsType::File => {
+                let (bytes, metadata, step) = IdleFileVisit::default()
+                    .start_from_parsed(flat, cache)?;
+
+                let (cid, name, depth) = self.next.expect("continued without next");
+                let file_continues = step.is_some();
+                self.current.as_file(cid, &name, depth, metadata, step);
+
+                let next = self.pending.pop();
+
+                // FIXME: add test case for this being reversed and it's never the last
+                let segment = FileSegment::first(bytes, !file_continues);
+
+                let state = if file_continues || next.is_some() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next,
+                        pending: self.pending,
+                    })
+                } else {
+                    State::Last(self.current)
+                };
+
+                Ok(ContinuedWalk::File(segment, Item::from(state)))
+            },
+            UnixFsType::Metadata => todo!("metadata?"),
```

Comment:
There is no code for this in `go-ipfs` expect test code as far as I understand, so I think it the type was an reservation of sorts. I think returning an unsupported error is probably enough... Or just "skip it", buffer any metadata for the next file, which should be the only link? I don't know if there are specs for this.